### PR TITLE
Internal type split

### PR DESCRIPTION
**Pull request by msujew**
This change splits the type system into a "plain" (string-based) type system and a concrete (reference-based) type system. Splitting the types allows for easier handling of purely generated information (i.e. container types, union/interface hierarchies and type names). It also has an impact on how we deal with property types and union types in general, which are now way more flexible.

Aims to resolve multiple issues:
* Closes https://github.com/langium/langium/issues/511 (The grammar has been slightly modified)
* Closes https://github.com/langium/langium/issues/476
* Closes https://github.com/langium/langium/issues/752
* Closes https://github.com/langium/langium/issues/534
* Closes https://github.com/langium/langium/issues/868

### CODE REVIEW

**Comment by spoenemann** [comment]
The result looks mostly really good! But it doesn't properly solve #534: with the rule
```
entry X: id=ID ({infer Y} 'a' | {infer Z} 'b');
```
 I get the types
```
export type X = Y | Z;

export interface Y extends AstNode {
    readonly $type: 'Y';
}

export interface Z extends AstNode {
    readonly $type: 'Z';
}
```
so the `id` property is forgotten. Could you add a test for this?

**Comment by msujew** [comment]
@spoenemann Thanks for the hint 👍 It should as expected now.

**Comment by spoenemann** [comment]
How does this relate to #554?

**Comment by pluralia** [comment]
The PR closes also #868.

**Comment by msujew** [comment]
> How does this relate to https://github.com/langium/langium/issues/554?

It's closely related, but not 100% there yet. My code can be used as a basis to build the graph mentioned in #554.

**Comment by msujew** [comment]
> After generating the code, the empty line at the end of these files is deleted

Unrelated to the changes in my PR. It might be due to some recent changes in the generator API.

**Comment by spoenemann** [comment]
>> After generating the code, the empty line at the end of these files is deleted
>
> Unrelated to the changes in my PR. It might be due to some recent changes in the generator API.

It's my fault: I had to manually fix the generated files (https://github.com/langium/langium/pull/888#issuecomment-1408359809), then @msujew fixed the generator, but we didn't regenerate the files. The file ending is different because the text editor automatically changes it on save.

File: packages/langium/src/grammar/validation/types-validator.ts [hunk 1]
```diff
@@ -233,120 +227,63 @@ function validateDeclaredAndInferredConsistency(typeInfo: InferredInfo & Declare
...
-        } else if (expectedPropertyType.array !== foundPropertyType.array || expectedPropertyType.reference !== foundPropertyType.reference) {
-            errorsInfo.push({ typeAsString, errorMessage: arrayReferenceError(foundPropertyType, expectedPropertyType) });
-        }
+function validateAlternativesConsistency(
+    inferred: PropertyType,
+    declared: PropertyType,
+    applyErrorToInferredTypes: (errorMessage: string) => void
+) {
+    if (!isTypeAssignable(inferred, declared)) {
+        applyErrorToInferredTypes(`Cannot assign type '${propertyTypeToString(inferred, 'DeclaredType')}' to '${propertyTypeToString(declared, 'DeclaredType')}'`);
```
Comment (montymxb) [inline packages/langium/src/grammar/validation/types-validator.ts:236]: Shouldn't the first one be 'InferredType' instead of 'DeclaredType'?

File: packages/langium/test/grammar/type-system/inferred-types.test.ts [hunk 2]
```diff
@@ -5,694 +5,331 @@
...
-        beforeAll(async () => {
-            const newTypes = await getTypes(grammar);
-            types.interfaces = newTypes.interfaces;
-            types.unions = newTypes.unions;
-        });
-        description.call(undefined, types);
-    });
-}
-
-describeTypes('inferred types of simple grammars', `
```
Comment (montymxb) [inline packages/langium/test/grammar/type-system/inferred-types.test.ts:31]: I'm not sure about these testing rewrites here. Doing validation on types by checking formatted strings feels very brittle to me going forward. Can we still not get the types that are produced by the collector, before printing, and verify on those instead?

If we could have the existing test structure, verifying our internal representation, plus this extra checking on the generated output, that would also work well I think.

Comment (montymxb) [reply packages/langium/test/grammar/type-system/inferred-types.test.ts:31]: To add as well, it's hard to follow if the integrity of the original tests is maintained with such a large refactor. I didn't get to look at it closely enough, but I would recommend a double check here by another colleague as well.

File: packages/langium/test/parser/langium-parser-builder.test.ts [hunk 3]
```diff
@@ -437,25 +437,25 @@ describe('Parser calls value converter', () => {
...
     test('Should parse bool correctly', () => {
         expectValue('b true', true);
-        // this is the current 'boolean' behavior when a prop type can't be resolved to just a boolean
-        // either true/undefined, no false in this case
-        expectValue('b false', undefined);
-        // ...then no distinguishing between the bad parse case when the type is unclear
-        expectValue('b asdfg', undefined);
+        expectValue('b false', false);
+        // Any value that cannot be parsed correctly is automatically false
+        expectValue('b asdfg', false);
```
Comment (montymxb) [inline packages/langium/test/parser/langium-parser-builder.test.ts:442]: This feels like a breaking change, especially if there's any upstream code that strictly relies on `undefined` to be produced.

File: packages/langium/test/parser/langium-parser-builder.test.ts [hunk 4]
```diff
@@ -437,25 +437,25 @@ describe('Parser calls value converter', () => {
...
     });
 
     test('Should parse BigInt correctly', () => {
         expectValue('big 9007199254740991n', BigInt('9007199254740991'));
-        expectValue('big 9007199254740991', undefined);
-        expectValue('big 1.1', undefined);
-        expectValue('big -19458438592374', undefined);
+        // Any value that cannot be parsed correctly is automatically false
+        expectValue('big 9007199254740991', false);
```
Comment (montymxb) [inline packages/langium/test/parser/langium-parser-builder.test.ts:448]: I would expect a numeric value that cannot be parsed to be `undefined`, rather than `false`. Is this fixed now by the design of the new implementation? If not, I would support reverting this.

**Review by montymxb — COMMENTED** [review]
General suggestions & comments. In particular I'm curious if we can still retain some (or all) of the previous inferred type tests structure. and then there's the point about the subtle change from `undefined` to `false` for values when parsing fails.

File: packages/langium/test/parser/langium-parser-builder.test.ts (hunk 3 above)
Comment (msujew) [reply packages/langium/test/parser/langium-parser-builder.test.ts:442]: That doesn't make a lot of sense, since a type that contains `boolean` is guaranteed to be `false` as a fallback (even if it's not parsed). Langium never generates a property as `value?: boolean`. This is actually a bug fix rather than a new feature.

File: packages/langium/test/parser/langium-parser-builder.test.ts (hunk 4 above)
Comment (msujew) [reply packages/langium/test/parser/langium-parser-builder.test.ts:448]: It's a bug fix, see the explanation above.

File: packages/langium/test/parser/langium-parser-builder.test.ts (hunk 3 above)
Comment (montymxb) [reply packages/langium/test/parser/langium-parser-builder.test.ts:442]: I'm aware it was a bug from before, but it's possible it will break something. I should clarify, in the next release notes we should make this change clear; but beyond that the basis for booleans not having an undefined value -- where it should be false -- is clear.

File: packages/langium/src/grammar/type-system/ast-collector.ts [hunk 5]
```diff
@@ -18,40 +18,58 @@ import { isPrimitiveType } from '../internal-grammar-util';
...
 
-    specifyAstNodeProperties(astTypes);
-    return astTypes;
+export function createAstTypes(first: PlainAstTypes, second?: PlainAstTypes): AstTypes {
+    const astTypes: PlainAstTypes = {
+        interfaces: mergeAndRemoveDuplicates<PlainInterface>(...first.interfaces, ...second?.interfaces ?? []),
+        unions: mergeAndRemoveDuplicates<PlainUnion>(...first.unions, ...second?.unions ?? []),
+    };
+
+    sortInterfacesTopologically(astTypes.interfaces);
```
[the remark this answers is not in the document]
Comment (msujew) [reply packages/langium/src/grammar/type-system/ast-collector.ts:40]: Right, thanks for the headsup! It's called correctly on `main` but I broke something during my latest rebase.

File: packages/langium/src/grammar/type-system/ast-collector.ts [hunk 6]
```diff
@@ -96,19 +96,44 @@ function filterInterfaceLikeTypes({ interfaces, unions }: AstTypes): Map<string,
     return nameToType;
 }
 
+function isDataType(property: PropertyType, visited: Set<PropertyType>): boolean {
```
[the remark this answers is not in the document]
Comment (msujew) [reply packages/langium/src/grammar/type-system/ast-collector.ts:99]: I personally like to keep the argument mandatory, since I usually forget to pass the arguments in the recursive call :)

File: packages/langium/test/parser/langium-parser-builder.test.ts (hunk 4 above)
Comment (montymxb) [reply packages/langium/test/parser/langium-parser-builder.test.ts:448]: The boolean case makes sense above, but here less so. It would be a topic I would like to discuss this in a later dev meeting perhaps, and we can move forward with this as is. Just want to be sure that we don't mislead or confuse end-users with how `false` is now a possible value for these property, even when the type is not a boolean.

File: packages/langium/src/grammar/type-system/ast-collector.ts [hunk 7]
```diff
@@ -96,19 +96,44 @@ function filterInterfaceLikeTypes({ interfaces, unions }: AstTypes): Map<string,
...
+            return isDataType(value.type, visited);
+        } else {
+            return false;
+        }
+    } else {
+        return isPrimitiveType(property) || isStringType(property);
+    }
+}
+
+function addSubTypes(types: TypeOption[]) {
```
[the remark this answers is not in the document]
Comment (msujew) [reply packages/langium/src/grammar/type-system/ast-collector.ts:118]: There is no further transformation happening at this point.

File: packages/langium/test/parser/langium-parser-builder.test.ts (hunk 4 above)
Comment (msujew) [reply packages/langium/test/parser/langium-parser-builder.test.ts:448]: Well, the type is a boolean. The generated type is `value: boolean | number | bigint | string | Date`.

File: packages/langium/src/grammar/validation/validator.ts [hunk 8]
```diff
@@ -659,14 +661,14 @@ export class LangiumGrammarValidator {
...
-    checkInterfacePropertyTypes(interfaceDecl: ast.Interface, accept: ValidationAcceptor): void {
-        interfaceDecl.attributes.filter(attr => attr.typeAlternatives.length > 1).forEach(attr => {
-            const typeKind = (type: ast.AtomType) => type.isRef ? 'ref' : 'other';
-            const firstKind = typeKind(attr.typeAlternatives[0]);
-            if (attr.typeAlternatives.slice(1).some(type => typeKind(type) !== firstKind)) {
-                accept('error', this.createMixedTypeError(attr.name), { node: attr, property: 'typeAlternatives' });
-            }
-        });
+    checkInterfacePropertyTypes(): void {
+        // interfaceDecl.attributes.filter(attr => attr.typeAlternatives.length > 1).forEach(attr => {
```
[the remark this answers is not in the document]
Comment (msujew) [reply packages/langium/src/grammar/validation/validator.ts:665]: I forgot to implement it after temporarily removing the validation 👍

File: packages/langium/src/grammar/validation/types-validator.ts (hunk 1 above)
Comment (msujew) [reply packages/langium/src/grammar/validation/types-validator.ts:236]: This is only for display. The other option would be `AstNode` which is for generation purposes only.

File: packages/langium/test/grammar/type-system/inferred-types.test.ts (hunk 2 above)
Comment (msujew) [reply packages/langium/test/grammar/type-system/inferred-types.test.ts:31]: The existing test structure was very tightly coupled with the exact layout of our old types. To prevent another test refactoring, I recommend that we only perform testing of types on the generated strings, as they contain all information that we get from the object model anyway.

File: packages/langium/src/grammar/validation/types-validator.ts [hunk 9]
```diff
@@ -199,29 +190,32 @@ function validateDeclaredAndInferredConsistency(typeInfo: InferredInfo & Declare
             accept('error', errorMessage, { node, property: ast.isAssignment(node) || ast.isAction(node) ? 'feature' : 'name' })
         );
 
+    // todo add actions
```
[the remark this answers is not in the document]
Comment (msujew) [reply packages/langium/src/grammar/validation/types-validator.ts:189]: I didn't touch any of that code. So I assume so. Maybe @pluralia knows a bit more.

File: packages/langium/test/parser/langium-parser-builder.test.ts (hunk 4 above)
Comment (montymxb) [reply packages/langium/test/parser/langium-parser-builder.test.ts:448]: Ahhh Right, that's a good point, in this example booleans are part of the prop, and so it would fallback...

So, in cases where `boolean` is not part of the type we should still get `undefined` then as before? I was looking over our cases and don't see any specific tests for checking that we get such a value when the parser fails (but no boolean type for the value). If that's how it works than I'm all good 👍 .

Comment (msujew) [reply packages/langium/test/parser/langium-parser-builder.test.ts:448]: Right, I'll add a test for that :)

File: packages/langium/test/grammar/type-system/inferred-types.test.ts (hunk 2 above)
Comment (montymxb) [reply packages/langium/test/grammar/type-system/inferred-types.test.ts:31]: Okay. I'll take a closer look at the tests to verify things look good in the restructure.

**Review by montymxb — APPROVED** [review]
Tests look good after a double check. There was one that was slightly changed (inferred types using chained actions), where G, H, and I are changed around. But this looks okay based on testing locally. Also checked this against some other Langium projects as well, and nothing appears out of the ordinary from what I can tell.
